### PR TITLE
feat: add opts.valid() and opts.beforeSave() hooks

### PR DESCRIPTION
**Pull request by fengmk2**
- valid() hook: valid session value before use it
- beforeSave() hook: set addition values before save it

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -472,6 +472,61 @@ describe('Koa Session', function(){
       });
     })
   })
+
+  describe('when valid and beforeSave set', function(){
+    it('should ignore session when uid changed', function(done){
```
Comment (fengmk2) [inline test/test.js]: This is an example.

**Comment by dead-horse** [comment]
LGTM

**Comment by dead-horse** [comment]
3.2.0

**Comment by fengmk2** [comment]
Thanks!

**Pull request merged**